### ISSUE TITLE
Fix PHP nightly builds

### ISSUE THREAD

**Issue by nicolas-grekas** [symfony/polyfill#218]
Here is a build that fails on PHP8:
https://travis-ci.org/symfony/polyfill/jobs/636310496

Help wanted to fix them!

**Comment by stof** [comment]
Most of the failures are related to PHP 8 converting warnings to TypeError for bad arguments.

We need to figure out what to do with them.

Note that if we want to polyfill the behavior exactly both for strict mode and loose mode, we might have to use typehints on the polyfill methods directly.

**Comment by Ayesh** [comment]
In regards to the exceptions thrown in nightly builds for type errors:

[PHP 8 will throw `\TypeError`](https://php.watch/versions/8.0/internal-function-exceptions) and [`\ValueError`](https://php.watch/versions/8.0/ValueError) exceptions, and this can be a huuuge break :)

This is no where near complete, but I think one approach would be to throw a `\TypeError` or `\ValueError` exception from the polyfill code, and conditionally expect a warning (PHPUnit-assisted warning) or an exception of type `\TypeError` or `\ValueError`. 

Ideally, we can drop PHP 5.x version support for good and add scalar typing. We will still need to work on `\ValueError` exceptions and work out differences thrown from within the function and at type-checking stage. 

Please see the second commit above. This approach worked to bring number of failures from [88](https://travis-ci.com/github/Ayesh/polyfill/jobs/314929103#L2199) to [76](https://travis-ci.com/github/Ayesh/polyfill/jobs/315205839#L2137) (this includes all errors throughout the polyfill suite).

**Comment by stof** [comment]
@Ayesh please open a pull request. It makes the review much easier.

**Comment by stof** [comment]
Well, for ValueError, we won't be able to make our polyfill turn existing core functions to the new behavior, as the native implementation would be used, not our polyfill.
So any code needing to support both PHP 7.x (and maybe 5.x) and PHP 8 will need to handle both kind of error handling (returning `false` with a warning, and throwing a `ValueError`) anyway.

What the polyfill should do is making sure that when it runs on PHP 8 for extensions that are not installed, it throws a `ValueError` rather than returning a `false` with a warning.

**Comment by Ayesh** [comment]
Thanks @stof - I did not feel like the commit above was complete, so I did not open a PR yet. 

> What the polyfill should do is making sure that when it runs on PHP 8 for extensions that are not installed, it throws a ValueError rather than returning a false with a warning.

Yes you are right. This is my approach in the commit above too. The warning messages are turned into `TypeError` or `\ValueError` exceptions. They should be in PHP 8 anyway, so we do not need to polyfill said classes. Instead, we can safely throw that exception with the same message verbatim. There is still work in progress in `php-src`, specially for `\ValueError` exceptions. So it is possible that the exception message can be different from the PHP 7.x warning. This will need to be handled per-case. The commit I have so far covers the verbatim error messages, but I will check the exception messages manually with current `php-src` master branch. Even when the PR is sent, it wouldn't make the builds green, but at least take care of mbstring tests :)
